File: go/wildling/parse_pattern.go
package wildling

import (
	"regexp"
	"strconv"
	"strings"
)

// Dictionaries maps dictionary names to word lists.
type Dictionaries map[string][]string

var (
	tokenParsingRegex      = regexp.MustCompile(`(\\[%@$*#&?!-]|[%@$*#&?!-]\{.*?\}|[%@$*#&?!-])`)
	lengthWithVariantsRegex = regexp.MustCompile(`\{((\d+)-(\d+)|(\d+))\}`)
	lengthWithStringRegex   = regexp.MustCompile(`\{'(.*)'(?:,(\d+)-(\d+))?(?:,(\d+))?\}`)
)

func parseLengthWithVariants(part string, variants []string) tokenOptions {
	startLength := 1
	endLength := 1

	match := lengthWithVariantsRegex.FindStringSubmatch(part)
	if match != nil {
		if match[2] != "" {
			startLength, _ = strconv.Atoi(match[2])
			endLength, _ = strconv.Atoi(match[3])
		} else if match[1] != "" {
			startLength, _ = strconv.Atoi(match[1])
			endLength = startLength
		}
	}

	return tokenOptions{
		Variants:    variants,
		StartLength: &startLength,
		EndLength:   &endLength,
		Src:         part,
	}
}

func parseLengthWithString(part string) (tokenOptions, bool) {
	match := lengthWithStringRegex.FindStringSubmatch(part)
	if match == nil {
		return tokenOptions{}, false
	}

	s := match[1]
	if match[2] != "" && match[3] != "" {
		start, _ := strconv.Atoi(match[2])
		end, _ := strconv.Atoi(match[3])
		return tokenOptions{
			String:      &s,
			StartLength: &start,
			EndLength:   &end,
			Src:         part,
		}, true
	}

	if match[4] != "" {
		length, _ := strconv.Atoi(match[4])
		return tokenOptions{
			String:      &s,
			StartLength: &length,
			EndLength:   &length,
			Src:         part,
		}, true
	}

	one := 1
	return tokenOptions{
		String:      &s,
		StartLength: &one,
		EndLength:   &one,
		Src:         part,
	}, true
}

func charsAsVariants(variantsString string) []string {
	variants := make([]string, 0, len(variantsString))
	for _, r := range variantsString {
		variants = append(variants, string(r))
	}
	return variants
}

func simpleTokenizer(variantsString string) func(string) *Token {
	variants := charsAsVariants(variantsString)
	return func(part string) *Token {
		return newToken(parseLengthWithVariants(part, variants))
	}
}

func dictionaryTokenizer(part string, dictionaries Dictionaries) *Token {
	options, ok := parseLengthWithString(part)
	if !ok || (options.String != nil && *options.String != "" && !dictHas(dictionaries, *options.String)) {
		one := 1
		return newToken(tokenOptions{
			Variants:    []string{part},
			StartLength: &one,
			EndLength:   &one,
			Src:         part,
		})
	}
	key := ""
	if options.String != nil {
		key = *options.String
	}
	options.Variants = dictionaries[key]
	if options.Variants == nil {
		options.Variants = []string{}
	}
	return newToken(options)
}

func dictHas(dictionaries Dictionaries, key string) bool {
	_, ok := dictionaries[key]
	return ok
}

func wordsTokenizer(part string) *Token {
	options, ok := parseLengthWithString(part)
	if !ok {
		one := 1
		return newToken(tokenOptions{
			Variants:    []string{part},
			StartLength: &one,
			EndLength:   &one,
			Src:         part,
		})
	}

	variants := []string{}
	workString := ""
	if options.String != nil {
		workString = *options.String
	}
	index := 0
	for index < len(workString) {
		if index+1 < len(workString) && workString[index] == '\\' && workString[index+1] == ',' {
			index += 2
		} else if workString[index] == ',' {
			variants = append(variants, workString[:index])
			workString = workString[index+1:]
			index = 0
		} else {
			index++
		}
	}
	variants = append(variants, workString)
	cleaned := make([]string, len(variants))
	for i, v := range variants {
		cleaned[i] = strings.ReplaceAll(v, "\\,", ",")
	}
	options.Variants = cleaned
	return newToken(options)
}

func partToToken(part string, dictionaries Dictionaries) *Token {
	tokenizers := map[byte]func(string) *Token{
		'#': simpleTokenizer("0123456789"),
		'@': simpleTokenizer("abcdefghijklmnopqrstuvwxyz"),
		'*': simpleTokenizer("abcdefghijklmnopqrstuvwxyz0123456789"),
		'-': simpleTokenizer("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"),
		'!': simpleTokenizer("ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
		'?': simpleTokenizer("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"),
		'&': simpleTokenizer("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"),
		'%': func(p string) *Token { return dictionaryTokenizer(p, dictionaries) },
		'$': wordsTokenizer,
	}

	var tokenizer func(string) *Token
	if len(part) > 0 {
		tokenizer = tokenizers[part[0]]
	}
	isEscaped := len(part) > 1 && part[0] == '\\' && tokenizers[part[1]] != nil

	if tokenizer != nil {
		return tokenizer(part)
	}
	if isEscaped {
		one := 1
		return newToken(tokenOptions{
			Variants:    []string{part[1:]},
			StartLength: &one,
			EndLength:   &one,
			Src:         part,
		})
	}
	one := 1
	return newToken(tokenOptions{
		Variants:    []string{part},
		StartLength: &one,
		EndLength:   &one,
		Src:         part,
	})
}

// ParsePattern splits a pattern into tokens.
func ParsePattern(inputPattern string, dictionaries Dictionaries) []*Token {
	if dictionaries == nil {
		dictionaries = Dictionaries{}
	}
	parts := splitKeepingDelimiters(inputPattern)
	tokens := make([]*Token, 0, len(parts))
	for _, part := range parts {
		if part == "" {
			continue
		}
		tokens = append(tokens, partToToken(part, dictionaries))
	}
	return tokens
}

// Go's regexp.Split does not include capturing-group matches (unlike JS/Python).
func splitKeepingDelimiters(input string) []string {
	indexes := tokenParsingRegex.FindAllStringSubmatchIndex(input, -1)
	if len(indexes) == 0 {
		if input == "" {
			return nil
		}
		return []string{input}
	}

	parts := make([]string, 0, len(indexes)*2+1)
	last := 0
	for _, loc := range indexes {
		// loc[0]:loc[1] full match; loc[2]:loc[3] group 1
		if loc[0] > last {
			parts = append(parts, input[last:loc[0]])
		}
		if loc[2] >= 0 && loc[3] >= loc[2] {
			parts = append(parts, input[loc[2]:loc[3]])
		}
		last = loc[1]
	}
	if last < len(input) {
		parts = append(parts, input[last:])
	}
	return parts
}

File: python/wildling/parse_pattern.py
from __future__ import annotations

import re
from typing import Callable, Dict, List, Union

from .token import Token, TokenOptions, create_token

Dictionaries = Dict[str, List[str]]

TOKEN_PARSING_REGEX = re.compile(
    r"(\\[%@$*#&?!-]|[%@$*#&?!-]\{.*?\}|[%@$*#&?!-])"
)


def parse_length_with_variants(part: str, variants: List[str]) -> TokenOptions:
    length_arg_regex = re.compile(r"\{((\d+)-(\d+)|(\d+))\}")
    match = length_arg_regex.search(part)

    start_length = 1
    end_length = 1

    if match is not None and match.group(2):
        start_length = int(match.group(2))
        end_length = int(match.group(3))
    elif match is not None and match.group(1):
        start_length = int(match.group(1))
        end_length = start_length

    return {
        "variants": variants,
        "startLength": start_length,
        "endLength": end_length,
        "src": part,
    }


def parse_length_with_string(part: str) -> Union[TokenOptions, bool]:
    length_arg_regex = re.compile(r"\{'(.*)'(?:,(\d+)-(\d+))?(?:,(\d+))?\}")
    match = length_arg_regex.search(part)

    if match is None:
        return False

    if match.group(2) is not None and match.group(3) is not None:
        return {
            "string": match.group(1) or "",
            "startLength": int(match.group(2)),
            "endLength": int(match.group(3)),
            "src": part,
        }

    if match.group(4) is not None:
        length = int(match.group(4))
        return {
            "string": match.group(1) or "",
            "startLength": length,
            "endLength": length,
            "src": part,
        }

    return {
        "string": match.group(1) or "",
        "startLength": 1,
        "endLength": 1,
        "src": part,
    }


def simple_tokenizer(variants_string: str) -> Callable[[str], Token]:
    variants = list(variants_string)

    def tokenizer(part: str) -> Token:
        return create_token(parse_length_with_variants(part, variants))

    return tokenizer


def _dictionary_tokenizer(part: str, dictionaries: Dictionaries) -> Token:
    options = parse_length_with_string(part)
    if options is False or (
        isinstance(options, dict)
        and options.get("string")
        and options["string"] not in dictionaries
    ):
        options = {
            "variants": [part],
            "startLength": 1,
            "endLength": 1,
            "src": part,
        }
    else:
        assert isinstance(options, dict)
        options["variants"] = dictionaries.get(options.get("string") or "", [])
    return create_token(options)


def _words_tokenizer(part: str) -> Token:
    options = parse_length_with_string(part)

    if options is False:
        options = {
            "variants": [part],
            "startLength": 1,
            "endLength": 1,
            "src": part,
        }
    else:
        assert isinstance(options, dict)
        variants: List[str] = []
        work_string = options.get("string") or ""
        index = 0
        while index < len(work_string):
            if work_string[index : index + 2] == "\\,":
                index += 2
            elif work_string[index] == ",":
                variants.append(work_string[:index])
                work_string = work_string[index + 1 :]
                index = 0
            else:
                index += 1
        variants.append(work_string)
        options["variants"] = [variant.replace("\\,", ",") for variant in variants]

    return create_token(options)


def part_to_token(part: str, dictionaries: Dictionaries) -> Token:
    tokenizers: Dict[str, Callable[[str], Token]] = {
        "#": simple_tokenizer("0123456789"),
        "@": simple_tokenizer("abcdefghijklmnopqrstuvwxyz"),
        "*": simple_tokenizer("abcdefghijklmnopqrstuvwxyz0123456789"),
        "-": simple_tokenizer(
            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        ),
        "!": simple_tokenizer("ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
        "?": simple_tokenizer("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"),
        "&": simple_tokenizer("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"),
        "%": lambda p: _dictionary_tokenizer(p, dictionaries),
        "$": _words_tokenizer,
    }

    tokenizer = tokenizers.get(part[0]) if part else None
    is_escaped_token = (
        len(part) > 1 and part[0] == "\\" and part[1] in tokenizers
    )

    if tokenizer is not None:
        return tokenizer(part)
    if is_escaped_token:
        return create_token(
            {
                "variants": [re.sub(r"^\\", "", part)],
                "src": part,
            }
        )
    return create_token({"variants": [part], "src": part})


def parse_pattern(input_pattern: str, dictionaries: Dictionaries) -> List[Token]:
    parts = [part for part in TOKEN_PARSING_REGEX.split(input_pattern) if part]
    return [part_to_token(part, dictionaries) for part in parts]
